In [1]:
import pandas

pandas.set_option('display.max_columns', None)

# Предварительный обзор данных

In [2]:
df = pandas.read_csv('./data/features.csv', index_col='match_id')

In [3]:
desc = df.describe()
desc

start_time    lobby_type       r1_hero      r1_level         r1_xp  \
count  9.723000e+04  97230.000000  97230.000000  97230.000000  97230.000000   
mean   1.444232e+09      2.630999     51.517104      3.442672   1233.405801   
std    5.515393e+06      2.835761     32.564211      1.111741    566.588895   
min    1.430199e+09      0.000000      1.000000      0.000000      0.000000   
25%    1.440815e+09      1.000000     22.000000      3.000000    767.000000   
50%    1.446338e+09      1.000000     50.000000      3.000000   1175.000000   
75%    1.448829e+09      7.000000     75.000000      4.000000   1704.000000   
max    1.450313e+09      7.000000    112.000000      6.000000   3319.000000   

            r1_gold         r1_lh      r1_kills     r1_deaths      r1_items  \
count  97230.000000  97230.000000  97230.000000  97230.000000  97230.000000   
mean    1147.899702     11.231996      0.357009      0.362285      8.271315   
std      464.111662      9.041620      0.663889      0.626704      2.497575   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%      746.000000      2.000000      0.000000      0.000000      7.000000   
50%     1113.000000     11.000000      0.000000      0.000000      8.000000   
75%     1479.000000     19.000000      1.000000      1.000000     10.000000   
max     4332.000000     47.000000      8.000000      5.000000     34.000000   

            r2_hero      r2_level         r2_xp       r2_gold         r2_lh  \
count  97230.000000  97230.000000  97230.000000  97230.000000  97230.000000   
mean      52.183452      3.364661   1189.223676   1107.863993     10.471747   
std       32.674077      1.097536    555.363510    458.001007      8.972073   
min        1.000000      0.000000      0.000000      0.000000      0.000000   
25%       25.000000      3.000000    740.000000    708.000000      2.000000   
50%       50.000000      3.000000   1109.000000   1054.000000      9.000000   
75%       75.000000      4.000000   1641.000000   1434.000000     18.000000   
max      112.000000      6.000000   3234.000000   4136.000000     44.000000   

           r2_kills     r2_deaths      r2_items       r3_hero      r3_level  \
count  97230.000000  97230.000000  97230.000000  97230.000000  97230.000000   
mean       0.348709      0.363550      8.275584     52.710491      3.353924   
std        0.654060      0.624236      2.433864     32.560923      1.092126   
min        0.000000      0.000000      0.000000      1.000000      0.000000   
25%        0.000000      0.000000      7.000000     25.000000      3.000000   
50%        0.000000      0.000000      8.000000     51.000000      3.000000   
75%        1.000000      1.000000     10.000000     76.000000      4.000000   
max        8.000000      6.000000     33.000000    112.000000      7.000000   

              r3_xp       r3_gold         r3_lh      r3_kills     r3_deaths  \
count  97230.000000  97230.000000  97230.000000  97230.000000  97230.000000   
mean    1182.879965   1099.643742     10.333004      0.342723      0.357338   
std      554.899600    454.727127      8.950871      0.647774      0.618071   
min        0.000000      0.000000      0.000000      0.000000      0.000000   
25%      736.000000    705.000000      2.000000      0.000000      0.000000   
50%     1101.000000   1040.000000      9.000000      0.000000      0.000000   
75%     1628.000000   1424.000000     18.000000      1.000000      1.000000   
max     3115.000000   3740.000000     51.000000      8.000000      6.000000   

           r3_items       r4_hero      r4_level         r4_xp       r4_gold  \
count  97230.000000  97230.000000  97230.000000  97230.000000  97230.000000   
mean       8.273527     52.648092      3.308896   1159.088481   1082.090240   
std        2.440139     32.670519      1.092502    550.937530    450.353291   
min        0.000000      1.000000      0.000000      0.000000      0.000000   
25%        7.000000     25.000000      3.000000    716

In [4]:
df.head(10)

start_time  lobby_type  r1_hero  r1_level  r1_xp  r1_gold  r1_lh  \
match_id                                                                     
0         1430198770           7       11         5   2098     1489     20   
1         1430220345           0       42         4   1188     1033      9   
2         1430227081           7       33         4   1319     1270     22   
3         1430263531           1       29         4   1779     1056     14   
4         1430282290           7       13         4   1431     1090      8   
5         1430284186           1       11         5   1961     1461     19   
8         1430293701           1        8         3    967     1136      7   
9         1430299335           7       35         5   2117     1252     16   
11        1430308974           1       17         5   1527      906     10   
12        1430316105           7       15         5   1651     1060     14   

          r1_kills  r1_deaths  r1_items  r2_hero  r2_level  r2_xp  r2_gold  \
match_id                                                                     
0                0          0         7       67         3    842      991   
1                0          1        12       49         4   1596      993   
2                0          0        12       98         3   1314      775   
3                0          0         5       30         2    539      539   
4                1          0         8       27         2    629      552   
5                0          1         6       20         2    441      686   
8                1          0         8       57         4   1774     1592   
9                0          0         6       15         3   1378     1430   
11               0          1         7       91         3    733      499   
12               0          1        10       41         3    951     1378   

          r2_lh  r2_kills  r2_deaths  r2_items  r3_hero  r3_level  r3_xp  \
match_id                                                                   
0            10         0          0         4       29         5   1909   
1            10         0          1         7       67         4   1506   
2             6         0          0         6       20         3   1297   
3             1         0          0         6       75         5   2037   
4             0         0          1         7       30         3    884   
5             4         0          0         5       28         4   1874   
8            19         1          0         8        7         3    904   
9            23         0          0         4       83         2    259   
11            0         0          0         4       53         4   1571   
12           19         0          0         6       74         4   1563   

          r3_gold  r3_lh  r3_kills  r3_deaths  r3_items  r4_hero  r4_level  \
match_id                                                                     
0            1143     10         0          0         8       20         3   
1            1502     18         1          0         7       37         3   
2             909      0         1          0         6       27         5   
3            1139     15         0          0         6       37         2   
4             927      0         1          0         8       72         3   
5            1438     22         0          0         4       25         2   
8            1119      7         0          1        12       21         4   
9             500      0         0          0         6       29         3   
11           1257      8         1          0         8       72         3   
12            855      8         0          1         4       90         3   

          r4_xp  r4_gold  r4_lh  r4_kills  r4_deaths  r4_items  r5_hero  \
match_id                                                                  
0           757      741      6         0          0         7      105   
1           669      631      7         0          0         

#### Описание признаков в таблице

- `match_id`: идентификатор матча в наборе данных
- `start_time`: время начала матча (unixtime)
- `lobby_type`: тип комнаты, в которой собираются игроки (расшифровка в `dictionaries/lobbies.csv`)
- Наборы признаков для каждого игрока (игроки команды Radiant — префикс `rN`, Dire — `dN`):
    - `r1_hero`: герой игрока (расшифровка в dictionaries/heroes.csv)
    - `r1_level`: максимальный достигнутый уровень героя (за первые 5 игровых минут)
    - `r1_xp`: максимальный полученный опыт
    - `r1_gold`: достигнутая ценность героя
    - `r1_lh`: число убитых юнитов
    - `r1_kills`: число убитых игроков
    - `r1_deaths`: число смертей героя
    - `r1_items`: число купленных предметов
- Признаки события "первая кровь" (first blood). Если событие "первая кровь" не успело произойти за первые 5 минут, то признаки принимают пропущенное значение
    - `first_blood_time`: игровое время первой крови
    - `first_blood_team`: команда, совершившая первую кровь (0 — Radiant, 1 — Dire)
    - `first_blood_player1`: игрок, причастный к событию
    - `first_blood_player2`: второй игрок, причастный к событию
- Признаки для каждой команды (префиксы `radiant_` и `dire_`)
    - `radiant_bottle_time`: время первого приобретения командой предмета "bottle"
    - `radiant_courier_time`: время приобретения предмета "courier" 
    - `radiant_flying_courier_time`: время приобретения предмета "flying_courier" 
    - `radiant_tpscroll_count`: число предметов "tpscroll" за первые 5 минут
    - `radiant_boots_count`: число предметов "boots"
    - `radiant_ward_observer_count`: число предметов "ward_observer"
    - `radiant_ward_sentry_count`: число предметов "ward_sentry"
    - `radiant_first_ward_time`: время установки командой первого "наблюдателя", т.е. предмета, который позволяет видеть часть игрового поля
- Итог матча (данные поля отсутствуют в тестовой выборке, поскольку содержат информацию, выходящую за пределы первых 5 минут матча)
    - `duration`: длительность
    - `radiant_win`: 1, если победила команда Radiant, 0 — иначе
    - Состояние башен и барраков к концу матча (см. описание полей набора данных)
        - `tower_status_radiant`
        - `tower_status_dire`
        - `barracks_status_radiant`
        - `barracks_status_dire`

## Обзор пропусков в данных

In [5]:
rows = len(df)
counts = desc.T['count']
counts_na = counts[counts < rows]
counts_na.sort_values().apply(lambda c: (rows - c) / rows)

first_blood_player2            0.452402
radiant_flying_courier_time    0.282619
dire_flying_courier_time       0.268415
first_blood_time               0.201100
first_blood_team               0.201100
first_blood_player1            0.201100
dire_bottle_time               0.166029
radiant_bottle_time            0.161380
radiant_first_ward_time        0.018883
dire_first_ward_time           0.018780
radiant_courier_time           0.007117
dire_courier_time              0.006953
Name: count, dtype: float64

### Обоснование пропусков:
* **first_blood_player2** (второй игрок, причастный к событию) - первое убийство может быть совершено одним игроком, это достаточно частое явление и происходит в 45% случаев
* **radiant_flying_courier_time** (время приобретения предмета "flying_courier") - в 28% случаев команда не покупает "flying_courier" в первые 5 минут игры
* **dire_flying_courier_time** - аналогично radiant_flying_courier_time, процент почти тот же (26%)
* **first_blood_time** (игровое время первой крови) - в 20% случаев первой крови нет в первые 5 минут игры
* **first_blood_team** (команда, совершившая первую кровь) - аналогично first_blood_time, процент совпадает (20%)
* **first_blood_player1** (игрок, причастный к событию) - аналогично first_blood_time, процент совпадает (20%)
* **dire_bottle_time** (время первого приобретения командой предмета "bottle") - в 16% случаев команда не приобретает предмет "bottle" в первые 5 минут игры
* **radiant_bottle_time** - аналогично dire_bottle_time, процент почти тот же (16%)
* **radiant_first_ward_time** (время установки командой первого "наблюдателя") - очень редкое явление, в 1% случаев команда не ставит "наблюдателя" в первые 5 минут игры
* **dire_first_ward_time** - аналогично dire_bottle_time, процент почти тот же (1%)
* **radiant_courier_time** (время приобретения предмета "courier") - самое редкое явление, в 0.5% случаев команда не приобретает предмет "courier" в первые 5 минут игры
* **dire_courier_time** - аналогично radiant_courier_time, процент почти тот же (0.5%)